[3.14] Add python version adjustment for frame count changes

fxdawnn opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168190

``sys.getrefcount(lib)`` got impacted due to python3.13 optimization. ``sys.getrefcount(lib._op_impls)`` and others remain the same.

Test plan:
``python test/test_python_dispatch.py TestPythonRegistration.test_finalizer`` in local ``python=3.14`` env

fxdawnn committed:
Add python version adjustment for frame count changes

[ghstack-poisoned]

williamwen42 reviewed (approved):
hmm, interesting that 3.13 seems to be affected by this too?

fxdawnn commented:
> hmm, interesting that 3.13 seems to be affected by this too?

waiting for CI changes to confirm.

fxdawnn committed:
Update on "Add python version adjustment for frame count changes"


``sys.getrefcount(lib)`` got impacted due to python3.13 optimization. ``sys.getrefcount(lib._op_impls)`` and others remain the same.

Test plan:
``python test/test_python_dispatch.py TestPythonRegistration.test_finalizer`` in local ``python=3.14`` env



[ghstack-poisoned]

fxdawnn commented:
@pytorchbot merge

miladm commented on test/test_python_dispatch.py:
```diff
@@ -299,8 +299,12 @@ def test_finalizer(self):
+        # 1 for `lib`, 1 for sys.getrefcount' for previous python version (<=3.12)
+        # In Python 3.13+, sys.getrefcount() was optimized to not create
+        # a temporary reference, so expected counts are 1 less than before
+        expected_refcount = 1 if sys.version_info >= (3, 14) else 2
```
do we generally hardcode this type of information in code - wondering if we have config files to track such changes @williamwen42?
